Core/Services/StitchCoreRemoteMongoDBService/Sources/StitchCoreRemoteMongoDBService/Sync/Internal/DataSynchronizer.swift

```diff
@@ -1237,70 +1351,61 @@ public class DataSynchronizer: NetworkStateDelegate, FatalErrorListener {
+                                                 documentId: AnyBSONValue,
                                                  document: Document,
                                                  atVersion: Document?,
-                                                 remoteEvent: ChangeEvent<Document>) throws {
-        guard let lock = syncConfig[namespace]?.nsLock else {
-            return
-        }
+                                                 remoteEvent: ChangeEvent<Document>) throws -> LocalSyncWriteModelContainer? {
+        let namespace = nsConfig.namespace
 
-        lock.assertWriteLocked()
-        guard let config = syncConfig[namespace]?[documentId] else {
-            return
-        }
-
-        let localCollection = self.localCollection(for: namespace, withType: Document.self)
-        let undoCollection = self.undoCollection(for: namespace)
-
-        let documentBeforeUpdate = try localCollection.find([idField: documentId], options: nil).next()
-
-        if var documentBeforeUpdate = documentBeforeUpdate {
-            try undoCollection.insertOne(&documentBeforeUpdate)
+        nsConfig.nsLock.assertWriteLocked()
+        guard let config = syncConfig[namespace]?[documentId.value] else {
+            return nil
         }
 
         // Remove forbidden fields from the resolved document before it will updated/upserted in the
         // local collection.
-        let docForStorage = DataSynchronizer.sanitizeDocument(document)
+        var docForStorage = DataSynchronizer.sanitizeDocument(document)
 
-        guard let documentAfterUpdate = try localCollection
-            .findOneAndReplace(
-                filter: ["_id": documentId],
-                replacement: docForStorage,
-                options: FindOneAndReplaceOptions(returnDocument: .after, upsert: true)) else {
-                    return
+        if !docForStorage.keys.contains(idField), remoteEvent.documentKey.keys.contains(idField) {
+            docForStorage[idField] = remoteEvent.documentKey[idField]
         }
 
         let event: ChangeEvent<Document>
-        if remoteEvent.operationType == .delete {
+        if case .delete = remoteEvent.operationType {
             event = ChangeEvents.changeEventForLocalInsert(namespace: namespace,
-                                                                    document: documentAfterUpdate,
-                                                                    documentId: documentId,
-                                                                    writePending: true)
+                                                           document: docForStorage,
+                                                           documentId: documentId.value,
+                                                           writePending: true)
         } else {
             guard let unsanitizedRemoteDocument = remoteEvent.fullDocument else {
-                return
+                return nil // XXX should raise an error here?
```

Comment:
this is fine for now. this case will become impossible once I do deltas anyway

Reply:
ok